### ISSUE TITLE
Remote debugger gets disconnected when trying to view a dictionary with freed nodes as its keys in the remote tab.

### ISSUE THREAD

**Issue by CritCorsac** [godotengine/godot#39868]
**Godot version:**
3.2.2

**Issue description:**
When trying to view a node that has an attached script that contains a dictionary with previously freed nodes as its keys in the remote tab, ~~the project will crash.~~ the remote debugger will get disconnected. This does not occur in 3.2.1 nor 3.2.2 beta 1 (I have not tested all of the 3.2.2 beta builds but I can confirm that the ~~crash~~ issue also occurs in 3.2.2 RC1).

**Steps to reproduce:**
Simply add the following code to a nodes script while the node is a parent of other nodes, then try to view that node in the remote tab while the project is running.
```
var test_dictionary = {}
func _ready():
	for a in get_children():
		test_dictionary[a] = "useless_value"
		a.queue_free()
```
**Minimal reproduction project:**
[DictionaryBug.zip](https://github.com/godotengine/godot/files/4840287/DictionaryBug.zip)

**Comment by pouleyKetchoupp** [comment]
Tested with 3.2.2 stable on Windows 10. I don't get a crash but the remote debugger gets disconnected and the log shows these errors:
```
ERROR: encode_variant: Condition "!v" is true. Returned: ERR_BUG
   At: core/io/marshalls.cpp:1179
ERROR: encode_variant: Condition "!v" is true. Returned: ERR_BUG
   At: core/io/marshalls.cpp:1179
ERROR: encode_variant: Condition "!v" is true. Returned: ERR_BUG
   At: core/io/marshalls.cpp:1179
ERROR: decode_variant: Condition "len < 4" is true. Returned: ERR_INVALID_DATA
   At: core/io/marshalls.cpp:108
ERROR: Error when trying to decode Variant.
   At: core/io/marshalls.cpp:490
ERROR: Error when trying to decode Variant.
   At: core/io/marshalls.cpp:525
ERROR: Error when trying to decode Variant.
   At: core/io/marshalls.cpp:525
ERROR: _notification: Condition "ret != OK" is true.
   At: editor/script_editor_debugger.cpp:1478
```

Can't reproduce on master b92477d77e9a6f46e4276a20a451dfac9d485f47

**Comment by CritCorsac** [comment]
Sorry, I should have clarified my issue a bit better. The editor itself doesn't crash, just the active project running in the debug window, causing the debug process to abruptly stop.

**Comment by pouleyKetchoupp** [comment]
@CritCorsac No worries! In my repro, the game itself actually doesn't crash. Only the remote debugging session stops and the game window stays open. Since it's different for you, could you share your minimal project in a zip? I wonder if the difference could be due to how we've set things up in the project somehow.

**Comment by CritCorsac** [comment]
@pouleyKetchoupp On further testing I am getting an identical error to you. I originally thought the game window itself had froze as I was using a simple print command to check if it was still responsive, but seeing as this bug disconnects the debugger I wouldn't have been getting an output anyway, making that a silly oversight by myself.

Sorry if I've been using any incorrect terminology. I'm still kind of new to all this.

I have added a minimal reproduction project to my original comment.

**Comment by pouleyKetchoupp** [comment]
@RandomShaper This looks like a regression on 3.2 from the dangling variants fix.

`ScriptDebuggerRemote::_send_object_id` fails on these lines when encoding the dictionary property:
https://github.com/godotengine/godot/blob/3e3972db7a0a7c04df8e852dc3e0d45061999575/core/io/marshalls.cpp#L1178-L1179

Because it fails to access the value from the key, which is a deleted object.

And that happens because `Variant::hash()` used in maps gets a null object in this case:
https://github.com/godotengine/godot/blob/3e3972db7a0a7c04df8e852dc3e0d45061999575/core/variant.cpp#L2859-L2862

On master, it works because `Variant::hash()` still gets an object pointer without checking instead:
https://github.com/godotengine/godot/blob/db168abe7d75d501a4a322dd8617f00ff8e2c4d5/core/variant.cpp#L2961-L2963

Would it make sense to change the `hash` method back to using the object address without nullifying it when it's been deleted? It's just a hash calculation, but I don't know if that could have dangerous side effects.

**Comment by akien-mga** [comment]
Fixed by #39906.
